src/domain/auth-rebac/domain/src/repos/rebac_repository.rs

```diff
@@ -0,0 +1,378 @@
+// the Business Source License, use of this software will be governed
+// by the Apache License, Version 2.0.
+
+use std::borrow::Cow;
+
+use internal_error::InternalError;
+use thiserror::Error;
+
+////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////
+
+#[async_trait::async_trait]
+pub trait RebacRepository: Send + Sync {
+    // Properties
+
+    async fn set_entity_property(
+        &self,
+        entity: &Entity,
+        property: PropertyName,
+        property_value: &PropertyValue,
+    ) -> Result<(), SetEntityPropertyError>;
+
+    async fn delete_entity_property(
+        &self,
+        entity: &Entity,
+        property_name: PropertyName,
+    ) -> Result<(), DeleteEntityPropertyError>;
+
+    async fn get_entity_properties(
+        &self,
+        entity: &Entity,
+    ) -> Result<Vec<(PropertyName, PropertyValue)>, GetEntityPropertiesError>;
+
+    // Relations
+
+    async fn insert_entities_relation(
+        &self,
+        subject_entity: &Entity,
+        relationship: Relation,
+        object_entity: &Entity,
+    ) -> Result<(), InsertEntitiesRelationError>;
+
+    async fn delete_entities_relation(
+        &self,
+        subject_entity: &Entity,
+        relationship: Relation,
+        object_entity: &Entity,
+    ) -> Result<(), DeleteEntitiesRelationError>;
+
+    async fn get_subject_entity_relations(
+        &self,
+        subject_entity: &Entity,
+    ) -> Result<Vec<ObjectEntityWithRelation>, SubjectEntityRelationsError>;
+
+    async fn get_subject_entity_relations_by_object_type(
+        &self,
+        subject_entity: &Entity,
+        object_entity_type: EntityType,
+    ) -> Result<Vec<ObjectEntityWithRelation>, SubjectEntityRelationsByObjectTypeError>;
+
+    async fn get_relations_between_entities(
+        &self,
+        subject_entity: &Entity,
+        object_entity: &Entity,
+    ) -> Result<Vec<Relation>, GetRelationsBetweenEntitiesError>;
+}
+
+////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////
+// DTOs
+////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////
+
+#[derive(Debug, Copy, Clone, PartialEq, Eq, Hash, PartialOrd, Ord)]
+pub enum EntityType {
+    Dataset,
+    Account,
+}
+
+////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////
+
+#[derive(Debug, Clone, PartialEq, Eq, Hash, PartialOrd, Ord)]
+pub struct Entity<'a> {
+    pub entity_type: EntityType,
+    pub entity_id: Cow<'a, str>,
+}
+
+impl<'a> Entity<'a> {
+    pub fn new(entity_type: EntityType, entity_id: impl Into<Cow<'a, str>>) -> Self {
+        Self {
+            entity_type,
+            entity_id: entity_id.into(),
+        }
+    }
+
+    pub fn new_account(entity_id: impl Into<Cow<'a, str>>) -> Self {
```

Comment:
Plz move all entity & relation structs into "entities" module, how we do it in all domains.

Reply:
Moved -- Done